lib/cart.js

```diff
@@ -0,0 +1,132 @@
+
+    // Calculate shipping using the loaded module
+    config.modules.loaded.shipping.calculateShipping(
+        req.session.totalCartNetAmount,
+        config,
+        req
+    );
+
+    // If discount module enabled
+    if(config.modules.loaded.discount){
+        // Recalculate discounts
+        const discount = await db.discounts.findOne({ code: req.session.discountCode });
+        if(discount){
+            config.modules.loaded.discount.calculateDiscount(
+                discount,
+                req
+            );
+        }else{
+            // If discount code is not found, remove it
+            delete req.session.discountCode;
+            req.session.totalCartDiscount = 0;
+        }
+    }
+
+    // Calculate our total amount removing discount and adding shipping
+    req.session.totalCartAmount = (req.session.totalCartNetAmount - req.session.totalCartDiscount) + req.session.totalCartShipping;
+};
+
+const updateSubscriptionCheck = (req, res) => {
+    // If cart is empty
+    if(!req.session.cart || req.session.cart.length === 0){
+        req.session.cartSubscription = null;
+        return;
+    }
+
+    Object.keys(req.session.cart).forEach((item) => {
+        if(item.productSubscription){
+            req.session.cartSubscription = item.productSubscription;
+        }else{
+            req.session.cartSubscription = null;
+        }
+    });
+};
+
+const emptyCart = async (req, res, type, customMessage) => {
+    const db = req.app.db;
+
+    // Remove from session
+    delete req.session.cart;
+    delete req.session.shippingAmount;
+    delete req.session.orderId;
+    delete req.session.cartSubscription;
+    delete req.session.discountCode;
+
+    // Remove cart from DB
+    await db.cart.deleteOne({ sessionId: req.session.id });
+
+    // update total cart
+    await updateTotalCart(req, res);
+
+    // Update checking cart for subscription
+    updateSubscriptionCheck(req, res);
+
+    // Set returned message
+    let message = 'Cart successfully emptied';
+    if(customMessage){
+        message = customMessage;
+    }
+
+    if(type === 'function'){
+        return;
+    }
+
+    // If POST, return JSON else redirect nome
+    if(type === 'json'){
+        res.status(200).json({ message: message, totalCartItems: 0 });
+        return;
+    }
+
+    req.session.message = message;
+    req.session.messageType = 'success';
+    res.redirect('/');
+};
+
+const clearCustomer = (req) => {
+    // Clear our session
+    req.session.customerPresent = null;
+    req.session.customerEmail = null;
+    req.session.customerFirstname = null;
+    req.session.customerLastname = null;
+    req.session.customerAddress1 = null;
+    req.session.customerAddress2 = null;
+    req.session.customerCountry = null;
+    req.session.customerState = null;
+    req.session.customerPostcode = null;
+    req.session.customerPhone = null;
+    req.session.orderComment = null;
+};
```

Comment:
This method seems odd in here, is it used by the cart and only by the cart?

Reply:
This is used by the `order` and the `customer` route so I felt moving/keeping in a helper/lib file would be the best option.